program-test/src/lib.rs

```diff
@@ -156,7 +153,6 @@ pub fn invoke_builtin_function(
 
     // Re-fetch the instruction context. The previous reference may have been
     // invalidated due to the `set_invoke_context` in a CPI.
-    let transaction_context = &invoke_context.transaction_context;
```

Comment:
The comment here is very important, I spent a couple of hours chasing this bug down a few years ago! This can't be removed

Reply:
Does not the comment refer to the next line after this?
```Rust
    let instruction_context = transaction_context.get_current_instruction_context()?;
```
How can the CPI call invalidate the reference created by:
```Rust
let transaction_context = &invoke_context.transaction_context;
```

Reply:
I moved this change to #2694. I don't have strong feelings about this line, I am just curious to understand the bug, since it looks like removing also the line:
```
let instruction_context = transaction_context.get_current_instruction_context()?;
```
does not make any test from `program-test` crate fail.

```Rust
pub fn invoke_builtin_function(
    builtin_function: solana_sdk::entrypoint::ProcessInstruction,
    invoke_context: &mut InvokeContext,
) -> Result<u64, Box<dyn std::error::Error>> {
    set_invoke_context(invoke_context);
```

The `invoke_context: &mut InvokeContext` reference which is used across the `invoke_builtin_function` should always be valid I think.